src/libraries/System.Drawing.Common/src/System/Drawing/Internal/GPStream.COMWrappers.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Runtime.InteropServices;
+
+namespace System.Drawing.Internal
+{
+    internal sealed partial class GPStream : Interop.Ole32.IStream
+    {
+        public unsafe Interop.HRESULT Clone(IntPtr* ppstm)
+        {
+            if (ppstm == null)
+            {
+                return Interop.HRESULT.STG_E_INVALIDPOINTER;
+            }
+
+            // The cloned object should have the same current "position"
+            var clone = new GPStream(_dataStream)
+            {
+                _virtualPosition = _virtualPosition
+            };
+
+            *ppstm = DrawingCom.Instance.GetOrCreateComInterfaceForObject(clone, CreateComInterfaceFlags.None);
+
+            return Interop.HRESULT.S_OK;
+        }
+
+        public unsafe Interop.HRESULT CopyTo(IntPtr pstm, ulong cb, ulong* pcbRead, ulong* pcbWritten)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+            ulong remaining = cb;
+            ulong totalWritten = 0;
+            ulong totalRead = 0;
+
+            fixed (byte* b = buffer)
+            {
+                while (remaining > 0)
+                {
+                    uint read = remaining < (ulong)buffer.Length ? (uint)remaining : (uint)buffer.Length;
+                    Read(b, read, &read);
+                    remaining -= read;
+                    totalRead += read;
+
+                    if (read == 0)
+                    {
+                        break;
+                    }
+
+                    uint written;
+                    Interop.HRESULT hr = (Interop.HRESULT)WriteToStream(pstm, b, read, &written);
+                    if (hr != Interop.HRESULT.S_OK)
+                    {
+                        return hr;
```

Comment:
do we need to return the buffer in this case?

Reply:
I've been told by @GrabYourPitchforks that when an error occurs, it is usually best to not return buffers to the pool, since they could have corrupt/incomplete data in them.

Reply:
It depends on how common you expect the errors can be. We do not bother to return the buffer to the pool for very rare errors, like out-of-memory exceptions.

The error here may not be that rare and it is not perf critical path, so it would be reasonable to return the buffer in finally here. This is similar to `Stream.CopyTo` where we do return in finally as well.

Reply:
>it is usually best to not return buffers to the pool, since they could have corrupt/incomplete data in them.

That is correct &ndash; especially in COM. A general rule for COM is that if a failure error code is returned all "out" parameters are in an indeterminate state. There is another school of thought that dictates the out/ref parameters shouldn't be touched in any manner until the callee can determine the result will be success.

Reply:
Note - the existing, "built-in COM" code doesn't return the buffer if an exception happens (which my understanding is that a COMException will occur since IStream defines `Write` as returning `void`, so the marshalling code will check the HR and throw):

https://github.com/dotnet/runtime/blob/be80f675c7aa1fc895c053f80229d2591c69e5eb/src/libraries/System.Drawing.Common/src/System/Drawing/Internal/GPStream.cs#L63-L89

@jkotas - do you think it is that common for an error to occur here?

Reply:
Probably not that common.

It is user supplied stream, so you can certainly constructs cases where the errors are common.

Reply:
I guess my preference is to leave it then, since that is the existing behavior and we haven't heard complaints.